subsys/net/lib/http/http_app.c

```diff
@@ -0,0 +1,247 @@
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	return net_app_close(&ctx->app_ctx);
+}
+
+int http_release(struct http_ctx *ctx)
+{
+	if (!ctx) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -ENOENT;
+	}
+
+	ctx->is_tls = false;
+
+#if defined(CONFIG_HTTP_SERVER) && defined(CONFIG_NET_DEBUG_HTTP_CONN)
+	if (!ctx->is_client) {
+		http_server_conn_del(ctx);
+		http_server_disable(ctx);
+	}
+#endif
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	ctx->is_init = false;
+
+	return net_app_release(&ctx->app_ctx);
+}
+
+int http_send_msg_raw(struct http_ctx *ctx, struct net_pkt *pkt,
+		      void *user_send_data)
+{
+	int ret;
+
+	NET_DBG("[%p] Sending %zd bytes data", ctx, net_pkt_get_len(pkt));
+
+	ret = net_app_send_pkt(&ctx->app_ctx, pkt, NULL, 0, 0,
+			       user_send_data);
+	if (!ret) {
+		/* We must let the system to send the packet, otherwise TCP
+		 * might timeout before the packet is actually sent. This is
+		 * easily seen if the application calls this functions many
+		 * times in a row.
+		 */
+		k_yield();
+	}
+
+	return ret;
+}
+
+static int _http_add_chunk(struct net_pkt *pkt, s32_t timeout,
+			   const u8_t *buf, size_t len)
+{
+	char chunk_header[16];
+	size_t ret;
+
+	if (!buf) {
+		len = 0;
+	}
+
+	snprintk(chunk_header, sizeof(chunk_header), "%x" HTTP_CRLF,
+		 (unsigned int)len);
+
+	ret = net_pkt_append(pkt, strlen(chunk_header), chunk_header,
+			     timeout);
+	if (ret == 0) {
+		return 0;
+	}
+
+	if (len) {
+		len = net_pkt_append(pkt, len, buf, timeout);
+		if (len == 0) {
+			return 0;
+		}
+	}
+
+	ret = net_pkt_append(pkt, sizeof(HTTP_CRLF) - 1, HTTP_CRLF,
+			     timeout);
+	if (ret == 0) {
+		return 0;
+	}
+
+	return len;
+}
+
+int http_send_data(struct http_ctx *ctx,
+		   const char *payload,
+		   size_t payload_len,
+		   void *user_send_data)
+{
+	size_t added;
+	int ret;
+
+	if (!ctx->pending) {
+		ctx->pending = net_app_get_net_pkt(&ctx->app_ctx,
+						   AF_UNSPEC,
+						   ctx->timeout);
+		if (!ctx->pending) {
+			return -ENOMEM;
+		}
+	}
+
+	ret = _http_add_chunk(ctx->pending, ctx->timeout,
+			      payload, payload_len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	added = ret;
+
+	if (added < payload_len) {
+		/* Not all data could be added, send what we have now
+		 * and return number of bytes sent to caller.
+		 */
+		ret = http_send_flush(ctx, user_send_data);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return added;
+}
+
+int http_send_flush(struct http_ctx *ctx, void *user_send_data)
+{
+	int ret;
+
+	if (!ctx->pending) {
+		return -ENOENT;
+	}
+
+	ret = http_send_msg_raw(ctx, ctx->pending, user_send_data);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctx->pending = NULL;
+
+	return ret;
+}
+
+static int _http_add_header(struct http_ctx *ctx, s32_t timeout,
+			    const char *name, const char *value)
+{
+	size_t added;
+
+	if (!ctx->pending) {
+		ctx->pending = net_app_get_net_pkt(&ctx->app_ctx, AF_UNSPEC,
```

Comment:
Per above, only http_send_data() allocates ctx->pending, and it's the only function which calls net_pkt_append(). All other functions use http_send_data() for any data transfer.